Libplanet.Explorer/Queries/StateQuery.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Security.Cryptography;
+using GraphQL;
+using GraphQL.Types;
+using Libplanet.Action;
+using Libplanet.Assets;
+using Libplanet.Blockchain;
+using Libplanet.Blockchain.Policies;
+using Libplanet.Blocks;
+using Libplanet.Explorer.GraphTypes;
+
+namespace Libplanet.Explorer.Queries;
+
+public class StateQuery<T>
+    : ObjectGraphType<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)>
+    where T : IAction, new()
+{
+    public StateQuery()
+    {
+        Name = "StateQuery";
+        Field<NonNullGraphType<FungibleAssetValueType>>(
+            "balance",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<AddressType>> { Name = "owner" },
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveBalance
+        );
+        Field<FungibleAssetValueType>(
+            "totalSupply",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveTotalSupply
+        );
+    }
+
+    private static Currency GetNativeTokenFromHash(
+        IResolveFieldContext<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)> context,
+        string name
+    )
+    {
+        byte[] currencyHashBytes = context.GetArgument<byte[]>(name);
+        HashDigest<SHA1> currencyHash;
+        try
+        {
+            currencyHash = new HashDigest<SHA1>(currencyHashBytes);
+        }
+        catch (Exception e)
+        {
+            throw new ExecutionError("currencyHash must consist of 20 bytes.\n" + e.Message, e);
```

Comment:
I'm wondering if the value of `name` is not `currencyHash`, would it be better to have this message to include the actual value of `name`.

Reply:
Right.  I'm going to remove the hard-coded parameter name.

Reply:
Amended.